### PR TITLE
TST: Clean old timezone issues PT2

### LINKED ISSUE #9243: Truncate on series causes core dump when TZ is specified on index
Hello
I am getting a core dump when truncating a series with a timezone specified on the index.  Please see below:

This works fine:

```
import pandas
import datetime
import numpy as np    
rng = pandas.date_range('1/2/2005', '11/1/2010', freq='D')
ts = pandas.Series(np.random.randn(len(rng)), index=rng)
ts.truncate(datetime.datetime(2009,1,1), datetime.datetime(2011,1,1)).resample('D', how='sum')
```

However this core dumps for me:

```
import pandas
import datetime
import numpy as np    
rng = pandas.date_range('1/2/2005', '11/1/2010', freq='D', tz='US/Eastern')
ts = pandas.Series(np.random.randn(len(rng)), index=rng)
ts.truncate(datetime.datetime(2009,1,1), datetime.datetime(2011,1,1)).resample('D', how='sum')
```

The only difference is adding the TZ parameter to the date_range to set the index for a particular timezone. 

This was working fine in .15, but seems to be broken in .15.2

Thanks

### LINKED ISSUE #11365: BUG: .loc assignment of datetime with tz is coercing to naive
xref #12502 (example there as well)
xref #14148 

```
In [3]: idx = pd.date_range('20130101',periods=3,tz='US/Eastern')

In [4]: df = DataFrame({'A' : [1,2,3]})

In [5]: df['B'] = idx

In [6]: df.dtypes
Out[6]: 
A                         int64
B    datetime64[ns, US/Eastern]
dtype: object

In [7]: df.loc[[True,False,True],'C'] = idx

In [8]: df
Out[8]: 
   A                         B                   C
0  1 2013-01-01 00:00:00-05:00 2013-01-01 05:00:00
1  2 2013-01-02 00:00:00-05:00                 NaT
2  3 2013-01-03 00:00:00-05:00 2013-01-02 05:00:00

# C should be the same time zone
In [9]: df.dtypes
Out[9]: 
A                         int64
B    datetime64[ns, US/Eastern]
C                datetime64[ns]
dtype: object
```

this is also realized to #11351 [4]

### LINKED ISSUE #11679: BUG: error in in _convert_to_indexer while using .loc with tz-aware DateTimeIndex
addtl example / repro in #13908 

Hello,

I encounter some problems using the data provided here: [data.txt](https://github.com/pydata/pandas/files/41644/data.txt)

When I was trying to use **.loc** on my **DateTimeIndex**, I go t the following error

``` python
KeyError: "['2015-03-01T02:00:00.000000000+0100'] not in index"
```

Notice that the error occurs only if we have the DST changing time in the DateTimeIndex

While debugging I've found that the error comes from the _convert_to_indexer method at: 

``` python
# indexing.py
mask = check == -1
if mask.any():
    raise KeyError('%s not in index' % objarr[mask]) <------------------------------------
```

Here is the copy-paste example using the data on the .txt attached:

``` python
data = pd.read_csv("data.txt"), parse_dates={'time':['Date']})
data.set_index('time', inplace=True)
data.index = data.index.tz_localize('Europe/Paris', ambiguous='infer').tz_convert('UTC')

for i in range(1, len(data)):
    try:-aware 
        data.loc[data.index[:i], 'value'] = -1
    except Exception as e:
        print i,e
        break

```

Other important fact: the error occurs for pandas version 0.17.0 and 0.17.1. However, I've managed to make it work with an older version of pandas (**0.16.2**) 

I hope this is explicit enough.

**UPDATE**

``` python
In[19]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Linux
OS-release: 3.19.0-33-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.7
pip: 7.1.2
setuptools: 18.5
Cython: 0.23.4
numpy: 1.10.1
scipy: 0.16.1
statsmodels: None
IPython: 4.0.0
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: None
tables: None
numexpr: 2.4.6
matplotlib: 1.5.0
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Jinja2: None
```

### LINKED ISSUE #11792: bug in Series.replace with timezone-aware datetime columns
Seeing the following behavior in 0.17.1 with the patch from https://github.com/pydata/pandas/pull/11715 applied (would raise without the patch):

``` python
In [29]: ser = pd.Series([pd.NaT, pd.Timestamp('2015/01/01', tz='UTC')])

# works
In [30]: ser.replace(pd.NaT, pd.Timestamp.min)
Out[30]: 
0    1677-09-22 00:12:43.145225
1     2015-01-01 00:00:00+00:00
dtype: object

# doesn't work
In [31]: ser.replace([np.nan, pd.NaT], pd.Timestamp.min)
Out[31]: 
0                         NaT
1   2015-01-01 00:00:00+00:00
dtype: datetime64[ns, UTC]

# works without timezone-aware datetimes
In [32]: ser = pd.Series([pd.NaT, pd.Timestamp('2015/01/01')])
In [33]: ser.replace([np.nan, pd.NaT], pd.Timestamp.min)
Out[33]: 
0   1677-09-22 00:12:43.145225
1   2015-01-01 00:00:00.000000
dtype: datetime64[ns]
```

### LINKED ISSUE #12064: BUG: date parsers use different tz offsets
xref #12060

@jreback - my thought was to just convert the `dateutil` timezone offset (`tzoffset`) to a pytz offset?

``` python
# dateutil
In [1]: pd.Timestamp('2013/01/01 00:00:00+09:00')
Out[1]: Timestamp('2013-01-01 00:00:00+0900', tz='tzoffset(None, 32400)')

# C parser
In [2]: pd.Timestamp('2013-01-01 00:00:00+09:00')
Out[2]: Timestamp('2013-01-01 00:00:00+0900', tz='pytz.FixedOffset(540)')

In [3]: pd.date_range('2013-01-01 00:00:00+09:00', '2014/01/01 00:00:00+09:00')

TypeError                                 Traceback (most recent call last)
<ipython-input-3-d34c322923bc> in <module>()
----> 1 pd.date_range('2013-01-01 00:00:00+09:00', '2014/01/01 00:00:00+09:00')

...

TypeError: Start and end cannot both be tz-aware with different timezones
```

### LINKED ISSUE #12601: BUG: Inconsistent datetime comparison with Tz
Related to #8306. On current master, `Timestamp` comparison results in `TypeError` if its timezones are different. However, `Index` and `Series` implicitly converts tz to GMT

```
pd.Timestamp('2016-01-01 12:00', tz='US/Eastern') > pd.Timestamp('2016-01-01 08:00')
# TypeError: Cannot compare tz-naive and tz-aware timestamps

# same result as idx.tz_convert(None) > pd.Timestamp('2016-01-01 08:00')
idx = pd.date_range('2016-01-01 12:00', periods=10, freq='H', tz='Asia/Tokyo')
idx > pd.Timestamp('2016-01-01 08:00')
# array([False, False, False, False, False, False,  True,  True,  True,  True], dtype=bool)
```

Numeric ops raises `TypeError` as expected.

```
idx - pd.Timestamp('2016-01-01 08:00')
# TypeError: Timestamp subtraction must have the same timezones or no timezones
```

### LINKED ISSUE #13051: BUG: coercion of mixed dt-aware data in Series constructor
I ran into an issue where the behavior of `.apply()` changed from 0.16 to 0.17, causing different results on tz aware data. Extracting the hour of day from datetimes is different for `Series(x).apply(func)` vs `func(x)`. Below is a minimal example of the issue in 0.17, it seems the behavior is the same in 0.18 but different (though still not equal) on master, also shown below.
#### On 0.17.1 and 0.18.0:

``` python
>>> import pandas as pd
>>> def hour_of_day(dt):
...     return dt.hour
...
>>> dt = pd.to_datetime(1462068217, unit='s')
>>> dt_localized = dt.tz_localize('UTC').tz_convert('US/Pacific')
>>> dt_list = [dt, dt_localized]
>>> apply_series = pd.Series(dt_list).apply(hour_of_day)
>>> map_series = pd.Series(map(hour_of_day, dt_list))
>>> print dt_list
[Timestamp('2016-05-01 02:03:37'), Timestamp('2016-04-30 19:03:37-0700', tz='US/Pacific')]
>>> print apply_series
0    2
1    2
dtype: int64
>>> print map_series
0     2
1    19
dtype: int64
>>> print apply_series - map_series
0     0
1   -17
dtype: int64
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 15.2.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.7
pip: 8.1.1
setuptools: 20.7.0
Cython: 0.24
numpy: 1.11.0
scipy: 0.17.0
statsmodels: None
IPython: 4.2.0
sphinx: 1.4.1
patsy: 0.4.1
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.5
lxml: 3.6.0
bs4: 4.3.2
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: 0.6.7.None
psycopg2: None
Jinja2: None
```
#### On 0.16.2 (what I expected):

``` python
# Same setup as above
>>> print apply_series
0     2
1    19
dtype: int64
>>> print map_series
0     2
1    19
dtype: int64
>>> print apply_series - map_series
0    0
1    0
dtype: int64
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 15.2.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.16.2
nose: None
Cython: None
numpy: 1.10.4
scipy: None
statsmodels: None
IPython: 4.2.0
sphinx: None
patsy: None
dateutil: 2.5.2
pytz: 2016.3
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
```
#### On master:

``` python
# Same setup as above
>>> print apply_series
0    19
1    19
dtype: int32
>>> print map_series
0     2
1    19
dtype: int64
>>> print apply_series - map_series
0    17
1     0
dtype: int64
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: 05e734ab171be0fda838c6b12839c38fa588da2c
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 15.2.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0+203.g05e734a
nose: None
pip: 8.1.1
setuptools: 20.7.0
Cython: None
numpy: 1.10.4
scipy: None
statsmodels: None
xarray: None
IPython: 4.2.0
sphinx: None
patsy: None
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
boto: None
pandas_datareader: None
```
#### Expected Output

I would expect the output to be [2, 19], as in 0.16, and matching map(f, data).

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #11679
- [x] closes #11365
- [x] closes #12064
- [x] closes #12601
- [x] closes #9243
- [x] closes #11792
- [x] closes #13051
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

xref #21491, cleaning up older timezone issues and moved the whatsnew entries in the first cleanup PR from v0.23.2 to v0.24.0 as discussed https://github.com/pandas-dev/pandas/pull/21491#discussion_r197077733

### CODE REVIEW

File: pandas/conftest.py [hunk 1]
```diff
@@ -249,3 +251,23 @@ def any_int_dtype(request):
     """
 
     return request.param
+
+
+COMPARISON_OPERATORS = [operator.lt, operator.le, operator.eq, operator.ne,
+                        operator.ge, operator.gt]
+
+
+@pytest.fixture(params=COMPARISON_OPERATORS)
```
Comment (jreback) [inline pandas/conftest.py]: make this more like this

```
+@pytest.fixture(params=['__eq__', '__ne__', '__le__',
+                        '__lt__', '__ge__', '__gt__'])
+def all_compare_operators(request):
+    """
+    Fixture for dunder names for common compare operations
     """
```

which is much more flexible for non-operators

File: pandas/tests/frame/test_indexing.py [hunk 2]
```diff
@@ -2248,6 +2262,16 @@ def test_setitem_datetimelike_with_inference(self):
                           index=list('ABCDEFGH'))
         assert_series_equal(result, expected)
 
+    @pytest.mark.parametrize('idxer', ['var', ['var']])
+    @pytest.mark.parametrize('tz', [None, 'UTC'])
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: try to use the tz fixtures if possible

File: pandas/tests/series/test_constructors.py [hunk 3]
```diff
@@ -1185,3 +1185,11 @@ def test_constructor_range_dtype(self, dtype):
         expected = Series([0, 1, 2, 3, 4], dtype=dtype or 'int64')
         result = Series(range(5), dtype=dtype)
         tm.assert_series_equal(result, expected)
+
+    def test_constructor_tz_aware_and_tz_naive_data(self):
```
Comment (jreback) [inline pandas/tests/series/test_constructors.py]: call this tz_mixed

File: pandas/tests/series/test_replace.py [hunk 4]
```diff
@@ -249,3 +249,12 @@ def test_replace_mixed_types_with_string(self):
         result = s.replace([2, '4'], np.nan)
         expected = pd.Series([1, np.nan, 3, np.nan, 4, 5])
         tm.assert_series_equal(expected, result)
+
+    @pytest.mark.parametrize('to_replace', [pd.NaT, [np.nan, pd.NaT]])
```
Comment (jreback) [inline pandas/tests/series/test_replace.py]: I think we have a fixture for nulls

File: pandas/tests/frame/test_indexing.py [hunk 5]
```diff
@@ -1688,6 +1688,20 @@ def test_getitem_list_duplicates(self):
         expected = df.iloc[:, 2:]
         assert_frame_equal(result, expected)
 
+    def test_getitem_setitem_datetimeindex_tz(self):
```
Comment (jreback) [inline pandas/tests/frame/test_indexing.py]: I think there are other tests like this, can you co-locate (IIRC in same file), *maybe* though in pandas/tests/indexing/....

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 6]
```diff
@@ -292,6 +292,15 @@ def test_construct_over_dst(self):
                             freq='H', tz='US/Pacific')
         tm.assert_index_equal(result, expected)
 
+    def test_construct_with_different_start_end_string_format(self):
+        # GH 12064
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_date_range.py]: not sure this is actually testing what is in the issue, as there were provided tz's there.

File: pandas/tests/indexes/datetimes/test_ops.py [hunk 7]
```diff
@@ -450,6 +450,14 @@ def test_offset_deprecated(self):
         with tm.assert_produces_warning(FutureWarning):
             idx.offset = BDay()
 
+    def test_ts_datetimeindex_compare_mismatched_tz(self, comparison_fixture):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_ops.py]: we have similar tests, can you co-locate

File: pandas/tests/series/test_replace.py [hunk 8]
```diff
@@ -249,3 +249,12 @@ def test_replace_mixed_types_with_string(self):
...
+    def test_replace_with_tz_aware_data(self, to_replace):
```
Comment (jreback) [inline pandas/tests/series/test_replace.py]: can you tests with tz-naive as well, maybe just have a combined test (as we likely have one for naive)

**Review by jreback — CHANGES_REQUESTED** [review]
@mroeschke can you break to 2 PRs, 1 which moves the issues out of 0.23.2 (just because want to cut 0.23.2 soon, so need to move)

File: pandas/tests/indexes/datetimes/test_date_range.py (hunk 6 above)
Comment (mroeschke) [reply pandas/tests/indexes/datetimes/test_date_range.py]: tz aren't provided in these examples via the keyword arg. The issue was that `'2013-01-01 00:00:00+09:00'` vs `'2013/01/01 02:00:00+09:00'` (note the `-` vs the `/`) ended up produced different tzoffsets

https://github.com/pandas-dev/pandas/issues/12064#issue-127047394

**Comment by mroeschke** [comment]
Addressed your comments @jreback and all green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mroeschke had some duplicate whatsnew entries, but fixed

**Pull request merged**